src/security/Handshake.cc

```diff
@@ -291,31 +293,26 @@ Security::HandshakeParser::parseModernRecord()
 void
 Security::HandshakeParser::parseMessages()
 {
-    try {
-        for (; !tkMessages.atEnd(); tkMessages.expectMore(false), tkMessages.commit()) {
-            switch (currentContentType) {
-                case ContentType::ctChangeCipherSpec:
-                    parseChangeCipherCpecMessage();
-                    continue;
-                case ContentType::ctAlert:
-                    parseAlertMessage();
-                    continue;
-                case ContentType::ctHandshake:
-                    parseHandshakeMessage();
-                    continue;
-                case ContentType::ctApplicationData:
-                    parseApplicationDataMessage();
-                    continue;
-            }
-            skipMessage("unknown ContentType msg [fragment]");
+    tkMessages.rollback();
+
+    for (; !tkMessages.atEnd(); tkMessages.commit()) {
```

Comment:
> we should move this tkMessages.commit() call into individual parsers 

Done.

>  we should stop parsing if we are `done`

It looks so but this `done` (described as `not nil if we got what we were looking for`) does not have a clear scope, i.e., whether it marks the end of records, messages or both (after these changes it became 'both').

>  let's make the above small step

Done.

Reply:
> > Alex: we should stop parsing if we are `done`

> It looks so but this `done` (described as `not nil if we got what we were looking for`) does not have a clear scope, i.e., whether it marks the end of records, messages or both (after these changes it became 'both').

I agree that `done` is not related to record or message boundaries, but why "but"? In other words, why is that lack of relationship represent a problem?

FWIW, I do not expect this PR changes to alter `done` meaning -- we may be "done" at any arbitrary input point.

Reply:
Imagine a scenario when we 'got what we were looking for' but there are still bytes in tkMessages which we should discard then. In this case, there would be a difference between the 'stop analyzing tkMessages immediately' and proceeding with the rest of tkMessages in parseMessages() (at least for debugging purposes, if these bytes are going to be thrown away anyway).

Reply:
> Imagine a scenario when we 'got what we were looking for' but there are still bytes in tkMessages which we should discard then.

I can easily imagine this scenario because our handshake parser is designed to parse until it gets the necessary information, not to parse the whole handshake. I would expect "leftovers" in tkRecords and/or tkMessages in many such cases. These leftovers do not represent a problem AFAICT.

If the handshake parser was designed to parse the entire handshake, we would not have the "done" field at all -- we would either successfully parse the whole thing, detect an error, or keep parsing.

Since Squid is not a part of that handshake in many cases, Squid will not be able to parse the encrypted parts.

> In this case, there would be a difference between the 'stop analyzing tkMessages immediately' and proceeding with the rest of tkMessages in parseMessages()

Yes, setting "done" is meant to stop parsing tkMessages and tkRecords. I do not see a problem with that. IMO, defining "done" as "successfully parsed everything we wanted to parse" or "got what we were looking for" results in fairly "clear scope".